File: docs/greedy/2412_hard/min_required_money.py

def calculate_min_required_money(transactions: list[list[int]]) -> int:  # LeetCode Q.2412.
    total_abs_loss = 0

    losing_trade_max_cashback = -1
    winning_trade_max_cost = -1

    for cost, cashback in transactions:
        if cashback > cost:
            if cost > winning_trade_max_cost:
                winning_trade_max_cost = cost
            continue

        abs_loss = cost - cashback
        total_abs_loss += abs_loss
        
        if cashback > losing_trade_max_cashback:
            losing_trade_max_cashback = cashback
    
    min_required_money = total_abs_loss + losing_trade_max_cashback
    
    if winning_trade_max_cost == -1: return min_required_money

    if min_required_money < winning_trade_max_cost + total_abs_loss:
        min_required_money = winning_trade_max_cost + total_abs_loss

    return min_required_money

File: docs/greedy/2412_hard/min_required_money.cpp
#include <vector>
using namespace std;

long long calculateMinRequiredMoney(vector<vector<int>> &transactions)
{ // LeetCode Q.2412.
  // Max cumulated losses <= min required money.
    long long totalAbsLoss = 0;

    int winningTradeMaxCost = -1;

    int losingTradeMaxCashback = -1;

    for (const auto &transaction : transactions)
    {
        int cost = transaction[0], cashback = transaction[1];

        if (cashback >= cost)
        {
            if (cost > winningTradeMaxCost)
                winningTradeMaxCost = cost;
            continue;
        }

        totalAbsLoss += cost - cashback;

        if (cashback > losingTradeMaxCashback)
            losingTradeMaxCashback = cashback;
    }

    long long minRequiredMoney = totalAbsLoss + losingTradeMaxCashback;

    if (winningTradeMaxCost == -1)
        return minRequiredMoney;

    if (minRequiredMoney < winningTradeMaxCost + totalAbsLoss)
        minRequiredMoney = winningTradeMaxCost + totalAbsLoss;

    return minRequiredMoney;
}
